src/PowerShellEditorServices/Workspace/Workspace.cs

```diff
@@ -624,6 +624,65 @@ private static string UnescapeDriveColon(string fileUri)
             return sb.ToString();
         }
 
+        /// <summary>
+        /// Converts a file system path into a DocumentUri required by Language Server Protocol.
+        /// </summary>
+        /// <remarks>
+        /// When sending a document path to a LSP client, the path must be provided as a
+        /// DocumentUri in order to features like the Problems window or peek definition
+        /// to be able to open the specified file.
+        /// </remarks>
+        /// <param name="path">
+        /// A file system path. Note: if the path is already a DocumentUri, it will be returned unmodified.
+        /// </param>
+        /// <returns>The file system path encoded as a DocumentUri.</returns>
+        public static string ConvertPathToDocumentUri(string path)
+        {
+            const string fileUriPrefix = "file:///";
+
+            if (path.StartsWith("untitled:", StringComparison.Ordinal))
+            {
+                return path;
+            }
+
+            if (path.StartsWith(fileUriPrefix, StringComparison.Ordinal))
+            {
+                return path;
+            }
+
+            string escapedPath = Uri.EscapeDataString(path);
+            var docUriStrBld = new StringBuilder(escapedPath);
+
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                // VSCode file URIs on Windows need the drive letter lowercase.
+                if (path.Contains(':'))
+                {
+                    for (int i = 1; i < docUriStrBld.Length - 2; i++)
+                    {
+                        if ((docUriStrBld[i] == '%') && (docUriStrBld[i + 1] == '3') && (docUriStrBld[i + 2] == 'A'))
+                        {
+                            int driveLetterIndex = i - 1;
+                            char driveLetter = char.ToLowerInvariant(docUriStrBld[driveLetterIndex]);
+                            docUriStrBld.Replace(path[driveLetterIndex], driveLetter, driveLetterIndex, 1);
+                            break;
+                        }
+                    }
+                }
+
+                // Uri.EscapeDataString goes a bit far, encoding \ chars. Besides VSCode wants / instead of \.
+                docUriStrBld.Replace("%5C", "/");
+            }
+            else
+            {
+                // Uri.EscapeDataString goes a bit far, encoding / chars.
+                docUriStrBld.Replace("%2F", "", 0, 3).Replace("%2F", "/");
```

Comment:
Can you give me an example of what this is doing?

Looks like it removes any `/` in the first 3 characters... but then replaces them with slashes after the fact? Is that for paths like:

```
/Users/tyler/blah/my.ps1
```

Reply:
Ah yeah, should have added a comment (will do that later).  The first removal happens because given a Linux/macOS path of `/home/tyler/blah/my.ps1` we will later prefix that with `file:///`.  If we don't remove the initial `/` you wind up with `file:////home/tyler/blah/my.ps1` which is incorrect.  It should be `file:///home/tyler/blah/my.ps1`.  From wikipedia:
> A valid file URI must therefore begin with either file:/path, file:///path or file://hostname/path.
file://path (i.e. two slashes, without a hostname) is never correct, but is often used.

VSCode expects `file:///` which is a bit tricky for Linux systems, 
> When using .NET Core on Unix-based systems, implicit file paths can be especially problematic, because an absolute implicit file path is indistinguishable from a relative path. When such ambiguity is present, Uri default to interpreting the path as an absolute URI.

From https://docs.microsoft.com/en-us/dotnet/api/system.uri?view=netframework-4.7.2#implicit-file-path-support

The other thing that occurs to me is that rather than removing the already escaped `/` here, we could add a `/` in the Windows case and then, when we insert the file scheme, we would insert `file://`.   I could go either way on this one.